src/elliptic/curves/wrappers.rs

```diff
@@ -0,0 +1,1692 @@
+    pub fn as_point_ref(&self) -> PointRef<E> {
+        PointRef::from(self)
+    }
+
+    fn from_raw(raw_point: E::Point) -> Result<Self, ZeroPointError> {
+        if raw_point.is_zero() {
+            Err(ZeroPointError(()))
+        } else {
+            Ok(Self { raw_point })
+        }
+    }
+
+    unsafe fn from_raw_unchecked(point: E::Point) -> Self {
+        Point { raw_point: point }
+    }
+
+    fn as_raw(&self) -> &E::Point {
+        &self.raw_point
+    }
+
+    fn into_raw(self) -> E::Point {
+        self.raw_point
+    }
+}
+
+impl<E: Curve> Clone for Point<E> {
+    fn clone(&self) -> Self {
+        //  Safety: `self` is guaranteed to be non-zero
+        unsafe { Point::from_raw_unchecked(self.as_raw().clone()) }
+    }
+}
+
+impl<E: Curve> fmt::Debug for Point<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.as_raw().fmt(f)
+    }
+}
+
+impl<E: Curve> TryFrom<PointZ<E>> for Point<E> {
+    type Error = ZeroPointError;
+    fn try_from(point: PointZ<E>) -> Result<Self, Self::Error> {
+        Self::from_raw(point.into_raw())
+    }
+}
+
+/// Elliptic curve generator
+///
+/// Holds internally a static reference on curve generator. Can be used in arithmetic interchangeably
+/// as [`PointRef<E>`](PointRef).
+///
+/// Generator multiplied at non-zero scalar always produce non-zero point, thus output type of
+/// the multiplication is [`Point<E>`](Point). This is the only difference compared to `Point<E>` and
+/// `PointRef<E>`. Use [`to_point_owned`](Self::to_point_owned) and [`as_point_ref`](Self::as_point_ref)
+/// methods to convert the generator into `Point<E>` and `PointRef<E>`.
+pub struct Generator<E: Curve> {
+    _ph: PhantomData<&'static E::Point>,
+}
+
+impl<E: Curve> Default for Generator<E> {
+    fn default() -> Self {
+        Self { _ph: PhantomData }
+    }
+}
+
+impl<E: Curve> Generator<E> {
+    fn as_raw(self) -> &'static E::Point {
+        E::Point::generator()
+    }
+
+    /// Clones generator point, returns `Point<E>`
+    pub fn to_point_owned(self) -> Point<E> {
+        // Safety: curve generator must be non-zero point, otherwise nothing will work at all
+        unsafe { Point::from_raw_unchecked(self.as_raw().clone()) }
+    }
+
+    /// Converts generator into `PointRef<E>`
+    pub fn as_point_ref(self) -> PointRef<'static, E> {
+        // Safety: curve generator must be non-zero point, otherwise nothing will work at all
+        unsafe { PointRef::from_raw_unchecked(self.as_raw()) }
+    }
+}
+
+impl<E: Curve> Clone for Generator<E> {
+    fn clone(&self) -> Self {
+        Self { _ph: PhantomData }
+    }
+}
+
+impl<E: Curve> Copy for Generator<E> {}
+
+/// Reference on elliptic point, _guaranteed_ to be non-zero
+///
+/// Holds internally a reference on [`Point<E>`](Point), refer to its documentation to learn
+/// more about Point/PointRef guarantees, security notes, and arithmetics.
+pub struct PointRef<'p, E: Curve> {
```

Comment:
Why is `PointRef` needed? (over `&Point`)

Reply:
Have a look at [Point::base_point2()](https://github.com/ZenGo-X/curv/blob/94a6459c24ef18fcf8266adf2deb14c29c5b9f74/src/elliptic/curves/wrappers.rs#L234) and [PointRef::base_point2()](https://github.com/ZenGo-X/curv/blob/94a6459c24ef18fcf8266adf2deb14c29c5b9f74/src/elliptic/curves/wrappers.rs#L405) methods.

They use [`ECPoint::base_point2()`](https://github.com/ZenGo-X/curv/blob/94a6459c24ef18fcf8266adf2deb14c29c5b9f74/src/elliptic/curves/traits.rs#L136) method underhood that returns a static reference to the 2nd base point: `&'static Self`. `PointRef<'p, E>` wraps `&'p E` (in the same way as `Point<E>` wraps `E`) and provides the same abilities as `Point<E>` has (arithmetic, serialization, etc.).

I'm very like this mechanism - old interface `base_point2()` method is less efficient as it constructs a point every time it called. I managed to workaround it, now the base_point2 is constructed only once, and the trait gives you a reference to its value.